app/forms/declaration/TotalNumberOfItems.scala

```diff
@@ -30,18 +30,34 @@ object TotalNumberOfItems extends DeclarationPage {
   implicit val format = Json.format[TotalNumberOfItems]
 
   val formId = "TotalNumberOfItems"
+  val exchangeRate = "exchangeRate"
+  val totalAmountInvoiced = "totalAmountInvoiced"
 
-  val totalAmountInvoicedPattern = "[0-9]{1,14}|[0-9]{1,14}[.][0-9]{1,2}"
-  val exchangeRatePattern = "[0-9]{1,12}|[[0-9]{1,7}[.][0-9]{1,5}]{3,13}"
+  val rateFieldErrorKey = "declaration.exchangeRate.error"
+  val invoiceFieldErrorKey = "declaration.totalAmountInvoiced.error"
+
+  val totalAmountInvoicedPattern = Seq("[0-9]{0,16}[.]{0,1}", "[0-9]{0,15}[.][0-9]{1}", "[0-9]{0,14}[.][0-9]{1,2}").mkString("|")
+
+  val exchangeRatePattern = Seq(
+    "[0-9]{0,12}[.]{0,1}",
+    "[0-9]{0,11}[.][0-9]{1}",
+    "[0-9]{0,10}[.][0-9]{1,2}",
+    "[0-9]{0,9}[.][0-9]{1,3}",
+    "[0-9]{0,8}[.][0-9]{1,4}",
+    "[0-9]{0,7}[.][0-9]{1,5}"
+  ).mkString("|")
+
+  val removeCommasFirst = (validator: String => Boolean) => (input: String) => validator(input.replaceAll(",", ""))
+  val notJustCommas = (input: String) => !input.forall(_.equals(','))
 
   val mapping = Forms.mapping(
-    "exchangeRate" -> optional(
+    exchangeRate -> optional(
       text()
-        .verifying("declaration.exchangeRate.error", isEmpty or ofPattern(exchangeRatePattern))
+        .verifying(rateFieldErrorKey, isEmpty or (notJustCommas and removeCommasFirst(ofPattern(exchangeRatePattern))))
```

Comment:
Maybe a comment to explain how `notJustCommas` and `removeCommasFirst` combine logically? (At least for me) it's not easy to grasp. As an example of me not understanding... why not checking if the resulting string is empty once `removeCommasFirst` is applied?

Reply:
The reason is that this field allows empty values (its an optional field). If a user types ",,," as the value and then we use `removeCommasFirst(isEmpty)` the validation would allow the nonsensical numeric value of ",,," through. So instead I check if the field is truly empty (that is allowed) or else that the field isn't just commas and is a numeric value in range. I'll try and add a comment explaining this